hive/src/main/java/com/netflix/iceberg/hive/HiveTableOperations.java

```diff
@@ -0,0 +1,235 @@
+
+      if (base != null) {
+        metaStoreClient.alter_table(database, tableName, tbl);
+      } else {
+        metaStoreClient.create_table(tbl);
+      }
+      threw = false;
+    } catch (TException | UnknownHostException e) {
+      throw new RuntimeException(String.format("Metastore operation failed for %s.%s", database, tableName), e);
+    } finally {
+      if (threw) {
+        // if anything went wrong, clean up the uncommitted metadata file
+        deleteFile(newMetadataLocation);
+      }
+      unlock(lockId);
+    }
+
+    refresh();
+  }
+
+  private void setParameters(String newMetadataLocation, Table tbl) {
+    Map<String, String> parameters = tbl.getParameters();
+
+    if (parameters == null) {
+      parameters = new HashMap<>();
+    }
+
+    parameters.put(TABLE_TYPE_PROP, ICEBERG_TABLE_TYPE_VALUE.toUpperCase(Locale.ENGLISH));
+    parameters.put(METADATA_LOCATION_PROP, newMetadataLocation);
+
+    if (currentMetadataLocation() != null && !currentMetadataLocation().isEmpty()) {
+      parameters.put(PREVIOUS_METADATA_LOCATION_PROP, currentMetadataLocation());
+    }
+
+    tbl.setParameters(parameters);
+  }
+
+  private StorageDescriptor storageDescriptor(Schema schema) {
+
+    final StorageDescriptor storageDescriptor = new StorageDescriptor();
+    storageDescriptor.setCols(columns(schema));
+    storageDescriptor.setLocation(hiveTableLocation());
+    storageDescriptor.setOutputFormat("org.apache.hadoop.mapred.FileInputFormat");
+    storageDescriptor.setInputFormat("org.apache.hadoop.mapred.FileOutputFormat");
+    SerDeInfo serDeInfo = new SerDeInfo();
+    serDeInfo.setSerdeType(SerdeType.HIVE);
+    serDeInfo.setSerializationLib("org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe");
+    storageDescriptor.setSerdeInfo(serDeInfo);
+    return storageDescriptor;
+  }
+
+  private final List<FieldSchema> columns(Schema schema) {
+    return schema.columns().stream().map(col -> new FieldSchema(col.name(), HiveTypeConverter.convert(col.type()), "")).collect(Collectors.toList());
+  }
+
+  private Long acquireLock() throws UnknownHostException, TException {
+    final LockComponent lockComponent = new LockComponent(LockType.EXCLUSIVE, LockLevel.TABLE, database);
+    lockComponent.setTablename(tableName);
+    final LockRequest lockRequest = new LockRequest(Lists.newArrayList(lockComponent),
+            System.getProperty("user.name"),
+            InetAddress.getLocalHost().getHostName());
+    LockResponse lockResponse = metaStoreClient.lock(lockRequest);
+    LockState state = lockResponse.getState();
+    Long lockId = lockResponse.getLockid();
+    //TODO add timeout
+    while (state.equals(LockState.WAITING)) {
+      lockResponse = metaStoreClient.check_lock(new CheckLockRequest(lockResponse.getLockid()));
+      state = lockResponse.getState();
+    }
+
+    if (!state.equals(LockState.ACQUIRED)) {
+      throw new IllegalStateException(String.format("Could not acquire the lock on %s.%s, " +
```

Comment:
Throwing CommitFailedException would trigger a retry. Is that what we want to do here, or should failure to get a lock fail the commit?

Reply:
I wasn't sure what scenarios could cause the state to be something else like ABORT or NOT_ACQUIRED and could not find any documentation so I assumed it is better to not retry but guess it won't hurt to retry given it will only delay failure. Changed it to CommitFailed.